diff --git a/src/inc/CMakeLists.txt b/src/inc/CMakeLists.txt
index 94daef9a6..305028ac5 100644
--- a/src/inc/CMakeLists.txt
+++ b/src/inc/CMakeLists.txt
@@ -32,6 +32,10 @@ if (ATOMIC)
     target_link_libraries(base_link INTERFACE ${ATOMIC})
 endif()
 
+if (LIBOQS)
+       target_link_libraries(base_link INTERFACE ${LIBOQS})
+endif()
+
 if (NUMA)
     target_link_libraries(base_link INTERFACE ${NUMA})
 endif()

