diff --git a/CMakeLists.txt b/CMakeLists.txt
index 419715963..16776981e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -423,6 +423,13 @@ if (NOT MSVC AND NOT APPLE AND NOT ANDROID)
     else()
         message(STATUS "libatomic not found. If build fails, install libatomic")
     endif()
+    find_library(LIBOQS NAMES oqs PATHS submodules/openssl/oqs/lib)
+    if (LIBOQS)
+           message(STATUS "Found liboqs: ${LIBOQS}")
+    else()
+           message(STATUS "liboqs not found at submodules/openssl/oqs/lib. If build fails, install liboqs")
+    endif()
+
 
     find_library(NUMA NAMES NUMA libnuma.so.1)
     if (NUMA)
