diff --git a/third_party/boringssl/BUILD.gn b/third_party/boringssl/BUILD.gn
index b435499f459b..0493ba2ec628 100644
--- a/third_party/boringssl/BUILD.gn
+++ b/third_party/boringssl/BUILD.gn
@@ -13,7 +13,7 @@ import("BUILD.generated_tests.gni")
 
 # Config for us and everybody else depending on BoringSSL.
 config("external_config") {
-  include_dirs = [ "src/include" ]
+  include_dirs = [ "src/include", "src/oqs/include" ]
   if (is_component_build) {
     defines = [ "BORINGSSL_SHARED_LIBRARY" ]
   }
@@ -45,7 +45,7 @@ config("no_asm_config") {
 }
 
 all_sources = crypto_sources + ssl_sources
-all_headers = crypto_headers + ssl_headers
+all_headers = crypto_headers + ssl_headers + oqs_headers
 
 # Windows' assembly is built with NASM. The other platforms use the platform
 # assembler. Exclude Windows ARM64 because NASM targets x86 and x64 only.
@@ -111,6 +111,7 @@ component("boringssl") {
   sources = all_sources
   public = all_headers
   friend = [ ":*" ]
+  libs = ["//third_party/boringssl/src/oqs/lib/liboqs.a"]
   deps = [ "//third_party/boringssl/src/third_party/fiat:fiat_license" ]
 
   # Mark boringssl_asm as a public dependency so the OPENSSL_NO_ASM
